src/Surfnet/StepupSelfService/SelfServiceBundle/Service/RemoteVettingService.php

```diff
@@ -0,0 +1,194 @@
+<?php
+
+/**
+ * Copyright 2019 SURFnet B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Surfnet\StepupSelfService\SelfServiceBundle\Service;
+
+use Psr\Log\LoggerInterface;
+use SAML2\Constants;
+use SAML2\XML\saml\SubjectConfirmation;
+use Surfnet\SamlBundle\Entity\IdentityProvider;
+use Surfnet\SamlBundle\Entity\ServiceProvider;
+use Surfnet\SamlBundle\Http\PostBinding;
+use Surfnet\SamlBundle\SAML2\AuthnRequestFactory;
+use Surfnet\StepupSelfService\SelfServiceBundle\Exception\InvalidRemoteVettingAuthenticationContextException;
+use Surfnet\StepupSelfService\SelfServiceBundle\Exception\InvalidRemoteVettingResponseException;
+use Surfnet\StepupSelfService\SelfServiceBundle\Service\RemoteVetting\IdentityProviderFactory;
+use Surfnet\StepupSelfService\SelfServiceBundle\Service\RemoteVetting\RemoteVettingTokenDto;
+use Surfnet\StepupSelfService\SelfServiceBundle\Service\RemoteVetting\ServiceProviderFactory;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+
+/**
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ */
+class RemoteVettingService
```

Comment:
I think we'll be adjusting this class quite a bit in the near future. Some things I noticed:

1. The lack of test coverage, this must be improved when the exact purpose of this class solidifies.
2. The service is tasked with too many responsibilities. The way I see this now, this class will mediate between the available remote vetting services and controller that interacts with them.

Reply:
1. See tasks above
2. I totally agree, as stated earlier on the goal was to implement the flow so work could be better split. Now indeed it will be time to refactor the service.